### PR TITLE
WIP: kernels

### PR DESCRIPTION
**Pull request by wsmoses**

### CODE REVIEW

**Comment by wsmoses** [comment]
@avik-pal @vchuravy @Pangoraw @jumerckx @mofeing obviously this is presently a reactant/enzyme/cassette frankenstein atm in utils.jl but cc'ing for visibility, and if you see an obvious reason I get the following error:

```
args = (typeof(typeof), ReactantCUDAExt.CuTracedArray{Int64, 1, 1, (64,)})
Square Kernel: Error During Test at /home/wsmoses/Reactant.jl/test/cuda.jl:20
  Got exception outside of a @test
  AssertionError: Unsupported call_with_reactant of builtin (typeof, 64 cu traced array at 0x00007fa1d6a46680)
  Stacktrace:
    [1] macro expansion
      @ ~/Reactant.jl/src/utils.jl:0 [inlined]
    [2] call_with_reactant(::typeof(typeof), ::ReactantCUDAExt.CuTracedArray{Int64, 1, 1, (64,)})
      @ Reactant ~/Reactant.jl/src/utils.jl:456
    [3] Typeof
      @ ./boot.jl:245 [inlined]
    [4] map
      @ ./tuple.jl:291 [inlined]
    [5] macro expansion
      @ ~/.julia/packages/CUDA/2kjXI/src/compiler/execution.jl:111 [inlined]
    [6] square!
      @ ~/Reactant.jl/test/cuda.jl:15 [inlined]
    [7] call_with_reactant(::typeof(square!), ::Reactant.TracedRArray{Int64, 1})
      @ Reactant ~/Reactant.jl/src/utils.jl:0
    [8] (::Reactant.var"#34#44"{Bool, Bool, typeof(square!), Tuple{ConcreteRArray{Int64, 1}}, Vector{Union{ReactantCore.MissingTracedValue, Reactant.TracedRArray, Reactant.TracedRNumber}}, Tuple{Reactant.TracedRArray{Int64, 1}}})()
      @ Reactant ~/Reactant.jl/src/utils.jl:559
    [9] block!(f::Reactant.var"#34#44"{Bool, Bool, typeof(square!), Tuple{ConcreteRArray{Int64, 1}}, Vector{Union{ReactantCore.MissingTracedValue, Reactant.TracedRArray, Reactant.TracedRNumber}}, Tuple{Reactant.TracedRArray{Int64, 1}}}, blk::Reactant.MLIR.IR.Block)
      @ Reactant.MLIR.IR ~/Reactant.jl/src/mlir/IR/Block.jl:201
   [10] make_mlir_fn(f::Function, args::Tuple{ConcreteRArray{Int64, 1}}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool, return_dialect::Symbol, no_args_in_result::Bool, construct_function_without_args::Bool, do_transpose::Bool)
      @ Reactant ~/Reactant.jl/src/utils.jl:541
   [11] make_mlir_fn
      @ ~/Reactant.jl/src/utils.jl:461 [inlined]
   [12] #10
      @ ~/Reactant.jl/src/Compiler.jl:295 [inlined]
   [13] block!(f::Reactant.Compiler.var"#10#15"{typeof(square!), Tuple{ConcreteRArray{Int64, 1}}}, blk::Reactant.MLIR.IR.Block)
      @ Reactant.MLIR.IR ~/Reactant.jl/src/mlir/IR/Block.jl:201
   [14] #9
      @ ~/Reactant.jl/src/Compiler.jl:294 [inlined]
   [15] mmodule!(f::Reactant.Compiler.var"#9#14"{Reactant.MLIR.IR.Module, typeof(square!), Tuple{ConcreteRArray{Int64, 1}}}, blk::Reactant.MLIR.IR.Module)
      @ Reactant.MLIR.IR ~/Reactant.jl/src/mlir/IR/Module.jl:92
   [16] compile_mlir!(mod::Reactant.MLIR.IR.Module, f::Function, args::Tuple{ConcreteRArray{Int64, 1}}; optimize::Bool)
      @ Reactant.Compiler ~/Reactant.jl/src/Compiler.jl:291
   [17] compile_mlir!
      @ ~/Reactant.jl/src/Compiler.jl:290 [inlined]
   [18] (::Reactant.Compiler.var"#34#36"{Bool, typeof(square!), Tuple{ConcreteRArray{Int64, 1}}})()
      @ Reactant.Compiler ~/Reactant.jl/src/Compiler.jl:698
   [19] context!(f::Reactant.Compiler.var"#34#36"{Bool, typeof(square!), Tuple{ConcreteRArray{Int64, 1}}}, ctx::Reactant.MLIR.IR.Context)
      @ Reactant.MLIR.IR ~/Reactant.jl/src/mlir/IR/Context.jl:76
   [20] compile_xla(f::Function, args::Tuple{ConcreteRArray{Int64, 1}}; client::Nothing, optimize::Bool)
      @ Reactant.Compiler ~/Reactant.jl/src/Compiler.jl:695
   [21] compile_xla
      @ ~/Reactant.jl/src/Compiler.jl:690 [inlined]
   [22] compile(f::Function, args::Tuple{ConcreteRArray{Int64, 1}}; client::Nothing, optimize::Bool, sync::Bool)
      @ Reactant.Compiler ~/Reactant.jl/src/Compiler.jl:722
   [23] macro expansion
      @ ~/Reactant.jl/src/Compiler.jl:475 [inlined]
   [24] macro expansion
      @ ~/Reactant.jl/test/cuda.jl:23 [inlined]
   [25] macro expansion
      @ ~/.julia/juliaup/julia-1.10.5+0.x64.linux.gnu/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [26] top-level scope
      @ ~/Reactant.jl/test/cuda.jl:21
   [27] include(fname::String)
      @ Base.MainInclude ./client.jl:489
   [28] top-level scope
      @ ~/Reactant.jl/test/runtests.jl:44
   [29] include(fname::String)
      @ Base.MainInclude ./client.jl:489
   [30] top-level scope
      @ none:6
   [31] eval
      @ ./boot.jl:385 [inlined]
   [32] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:291
   [33] _start()
      @ Base ./client.jl:552
Test Summary: | Error  Total     Time
Square Kernel |     1      1  1m54.4s
```

I realized while doing this that a pre req we need is supporting custom absint through indirect function calls so most of utils.jl tries to do this (and incidentally in doing so starts to set up the infra we will need later on for automated detection of if/for/etc into traced versions .

Of course atm everything breaks, but anyways

**Comment by wsmoses** [comment]
@aviatesk obviously this is still very in progress, but I'm wondering if you can give it a look?

Essentially the absint part is trying to make a code call that forces the use of our abstract interpreter, for all type unstable calls (e.g. unlike cassette we run inference and optimizations first). Eventually we also will do more with the IR as well like raising control flow into nice structures, but thats a future todo

**Comment by wsmoses** [comment]
I need to finish up registration, but kernel compile and execute (phase 1) now works!

```julia
#= /home/wsmoses/Reactant.jl/test/cuda.jl:22 =# @code_hlo(square!(A)) = module {
  func.func @main(%arg0: tensor<64xi64>) -> tensor<64xi64> {
    %0 = stablehlo.custom_call @reactant_gpu_call(%arg0) {output_operand_aliases = [#stablehlo.output_operand_alias<output_tuple_indices = [], operand_index = 0, operand_tuple_indices = []>]} : (tensor<64xi64>) -> tensor<64xi64>
    return %0 : tensor<64xi64>
  }
}
(res, xs) = (64 cu traced array at 0x00007f57d89256f0, TracedRArray{Int64,1N}(((:args, 1),), size=(64,)))
("recufunction", f, tt) = ("recufunction", square_kernel!, Tuple{ReactantCUDAExt.CuTracedArray{Int64, 1, 1, (64,)}})
args = (64 cu traced array at 0x00007f57d89256f0,)
call_kwargs = Base.Pairs{Symbol, Union{}, Tuple{}, @NamedTuple{}}()
a = 64 cu traced array at 0x00007f57d89256f0
arg = %1 = "stablehlo.transpose"(%0) <{permutation = array<i64: 0>}> : (tensor<64xi64>) -> tensor<64xi64>
terminate called after throwing an instance of 'xla::XlaRuntimeError'
  what():  UNIMPLEMENTED: No registered implementation for custom call to reactant_gpu_call for platform CUDA
```

**Comment by wsmoses** [comment]
also cc for vis @ivanradanov @ftynse we now [almost] have a single mlir file with host and device code for an entire julia application

File: ext/ReactantCUDAExt.jl [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+		length(args) == 1 ? C_NULL : Ref{Int64}(i-1),
+		i-1,
+		0,
+		C_NULL
+		))
+	      )
+    end
+
+    output_operand_aliases=MLIR.IR.Attribute(aliases)
+    call = MLIR.Dialects.stablehlo.custom_call(mlir_args; result_0=restys, call_target_name="reactant_gpu_call", output_operand_aliases)
```
Comment (ftynse) [inline ext/ReactantCUDAExt.jl:427]: FYI, to pass some information to the XLA callback of this kernel, use the `backend_config` (https://github.com/openxla/stablehlo/blob/ef176a130f28196dcb4a5735d0f2f6ed0f85bd5d/stablehlo/dialect/StablehloOps.td#L2391) string attribute. That attribute gets forwarded to the "opaque" pair of arguments (pointer + size) as a C string.

**Comment by wsmoses** [comment]
I still need to make the nicer op with registration and lowerings (incl https://github.com/llvm/llvm-project/pull/119998), but we now have host and device code in an mlir module

```mlir
#= /home/wmoses/git/Reactant.jl/test/cuda.jl:22 =# @code_hlo(square!(A)) = #tbaa_root = #llvm.tbaa_root<id = "custom_tbaa">
#tbaa_type_desc = #llvm.tbaa_type_desc<id = "custom_tbaa_addrspace(1)", members = {<#tbaa_root, 0>}>
#tbaa_tag = #llvm.tbaa_tag<base_type = #tbaa_type_desc, access_type = #tbaa_type_desc, offset = 0>
module {
  llvm.func internal unnamed_addr fastcc @julia_throw_boundserror_2676() attributes {dso_local, no_inline, sym_visibility = "private"} {
    llvm.unreachable
  }
  llvm.func local_unnamed_addr ptx_kernelcc @_Z14square_kernel_13CuTracedArrayI5Int64Ll1ELl1E5_64__E(%arg0: !llvm.array<1 x ptr<1>>) {
    %0 = llvm.mlir.constant(63 : i32) : i32
    %1 = nvvm.read.ptx.sreg.tid.x : i32
    %2 = llvm.icmp "ugt" %1, %0 : i32
    llvm.cond_br %2, ^bb2, ^bb1
  ^bb1:  // pred: ^bb0
    %3 = llvm.extractvalue %arg0[0] : !llvm.array<1 x ptr<1>>
    %4 = llvm.zext %1 : i32 to i64
    %5 = llvm.getelementptr inbounds %3[%4] : (!llvm.ptr<1>, i64) -> !llvm.ptr<1>, i64
    %6 = llvm.load %5 {alignment = 1 : i64, tbaa = [#tbaa_tag]} : !llvm.ptr<1> -> i64
    %7 = llvm.mul %6, %6 : i64
    llvm.store %7, %5 {alignment = 1 : i64, tbaa = [#tbaa_tag]} : i64, !llvm.ptr<1>
    llvm.return
  ^bb2:  // pred: ^bb0
    llvm.call fastcc @julia_throw_boundserror_2676() : () -> ()
    llvm.unreachable
  }
  func.func @main(%arg0: tensor<64xi64>) -> tensor<64xi64> {
    %0 = stablehlo.custom_call @reactant_gpu_call(%arg0) {backend_config = "_Z14square_kernel_13CuTracedArrayI5Int64Ll1ELl1E5_64__E", output_operand_aliases = [#stablehlo.output_operand_alias<output_tuple_indices = [], operand_index = 0, operand_tuple_indices = []>]} : (tensor<64xi64>) -> tensor<64xi64>
    return %0 : tensor<64xi64>
  }
}
call_kwargs = Base.Pairs{Symbol, Union{}, Tuple{}, @NamedTuple{}}()
blockdim = CuDim3(0x00000001, 0x00000001, 0x00000001)
threaddim = CuDim3(0x00000040, 0x00000001, 0x00000001)
```

**Comment by wsmoses** [comment]
This still has some more work to go before fully working, but it improves the errors and current state (And resolves other interpreter things and is a prerequisite to a stack of downstream opts). I'm going to go ahead and merge for now and continue in follow ups

**Pull request merged**